qiskit/transpiler/passes/optimization/collect_multiqubit_blocks.py

```diff
@@ -0,0 +1,212 @@
+                if tot_size > self.max_block_size:
+                    makes_too_big = True
+
+            if not can_process:
+                # resolve the case where we cannot process this node
+                for bit in cur_qubits:
+                    # create a gate out of me
+                    bit = self.find_set(bit)
+                    if len(self.gate_groups[bit]) == 0:
+                        continue
+                    block_list.append(self.gate_groups[bit][:])
+                    cur_set = set(self.bit_groups[bit])
+                    for v in cur_set:
+                        # reset this bit
+                        self.parent[v] = v
+                        self.bit_groups[v] = [v]
+                        self.gate_groups[v] = []
+
+            if makes_too_big:
+                # adding in all of the new qubits would make the group too big
+                # we must block off sub portions of the groups until the new
+                # group would no longer be too big
+                savings = {}
+                groups_seen = set()
+                tot_size = 0
+                for bit in cur_qubits:
+                    top = self.find_set(bit)
+                    if top in groups_seen:
+                        savings[top] = savings[top] - 1
+                    else:
+                        groups_seen.add(top)
+                        savings[top] = len(self.bit_groups[top]) - 1
+                        tot_size += len(self.bit_groups[top])
+                slist = []
+                for item in groups_seen:
+                    slist.append((savings[item], item))
+                slist.sort(reverse=True)
+                savings_need = tot_size - self.max_block_size
+                for item in slist:
+                    # remove groups until the size created would be acceptable
+                    # start with blocking out the group that would decrease
+                    # the new size the most. This heuristic for which blocks we
+                    # create does not necessarily give the optimal blocking. Other
+                    # heuristics may be worth considering
+                    if savings_need > 0:
+                        savings_need = savings_need - item[0]
+                        if len(self.gate_groups[item[1]]) >= 1:
+                            block_list.append(self.gate_groups[item[1]][:])
+                        cur_set = set(self.bit_groups[item[1]])
+                        for v in cur_set:
+                            self.parent[v] = v
+                            self.bit_groups[v] = [v]
+                            self.gate_groups[v] = []
+
+            if can_process:
+                # if the operation is a gate, either skip it if it is too large
+                # or group up all of the qubits involved in the gate
+                if len(cur_qubits) > self.max_block_size:
+                    continue # unable to be part of a group
```

Comment:
We should not. In this case, all of those blocks would already have been finalized due to the makes_too_big section. The purpose of this code is so that we don't add any gates that are too big to blocks. These gates cannot exist in a block on max_block_size qubits and so we just ignore them. I added a comment clarifying this case in the code.